### PR TITLE
Stop eager symbolic materialization for tolist sum

### PR DESCRIPTION
**Pull request by bobrenjc93**
## Summary

This stops `Tensor.tolist()` in Dynamo from eagerly wrapping every element of non-scalar integer tensors. `tolist()` now returns a lazy tensor-backed list variable and only materializes individual symbolic elements when list operations actually need them. `sum(x.tolist())` is handled directly as a single tensor reduction plus scalar capture.

## Root Cause Problem

`TensorVariable.method_tolist()` recursively wrapped every integer tensor element with `getitem().item()` as soon as `tolist()` was called. For patterns like `sum(x.tolist())`, Dynamo built hundreds of per-element symbolic nodes before the existing `sum()` fast path had a chance to reduce them.

## Proposed Fix

Add a lazy `TensorToListVariable` for non-scalar integer `Tensor.tolist()` results. It preserves list-like behavior by materializing on indexing, iteration, list arithmetic, tensor method/function arguments, or reconstruction, but exposes a targeted `sym_sum()` path so `sum(x.tolist())` traces as one reduction instead of one scalar per element.

## Why This Is The Right Long Term Fix

The expensive behavior is caused by representing a list eagerly at `tolist()` time even when the consumer is a reduction. Making the sequence lazy keeps normal list semantics available while allowing reduction consumers to choose a compact representation. This keeps the optimization localized to Dynamo variable tracking instead of adding broad graph rewrites after hundreds of nodes already exist.

## Tests

- `python3 -m py_compile torch/_dynamo/variables/tensor.py torch/_dynamo/variables/functions.py torch/_dynamo/variables/builtin.py torch/_dynamo/variables/torch.py test/dynamo/test_misc.py`
- `PYTHONPATH=. python test/dynamo/test_misc.py -k test_tolist_sum_uses_lazy_reduction`
- `PYTHONPATH=. python test/dynamo/test_misc.py MiscTests.test_tolist_1d MiscTests.test_tolist_kd MiscTests.test_tolist_0d MiscTests.test_tolist_kd_dynamic MiscTests.test_tolist_sum_uses_lazy_reduction`
- Regression check without implementation: new test fails with `counter.op_count == 202`, then passes with `counter.op_count == 3`
- `lintrunner --take PYFMT ...`
- `lintrunner --take RUFF,FLAKE8,TYPEIGNORE ...`
- Eager smoke: compiled `torch.split(x, y.tolist())` still materializes list sizes correctly
- Inductor smoke: compiled `sum(x.tolist())` returns the expected scalar tensor

Drafted via Codex, published after manual review by @bobrenjc93

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in commit 54ff0ee0386: `reconstruct()` now uses `codegen.tx`, `as_proxy()` documents the retained translator lifetime, inlined returns materialize lazy `tolist()` values consistently, and `materialize_tensor_tolist_arg()` now recurses through list/tuple-like containers. I did not add the suggested extra 2D `sum(x.tolist())` fallback test because it would be a non-regression test that already passes without this fix; multi-dimensional `tolist()` materialization is covered by `test_tolist_kd`, while the new red/green regression is focused on the CI failure in `torch.tensor(x.tolist())`.

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Pushed follow-up commit a7d49a8e374 to address the current CI failures. The root cause was recursive lazy tolist materialization using generic VariableTracker type checks, which could force LazyVariableTracker realization and also recurse through self-referential lists. The fix limits return-boundary and sum fast-path checks to exact TensorToListVariable values, keeps nested materialization cycle-safe, and adds the requested sym_sum comment about int64 overflow semantics. I did not add the suggested 2D sum fallback test because it is not a red/green regression for this PR; existing coverage already checks multidimensional tolist materialization correctness.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Pushed follow-up commit 174b373b8d7 for the latest Claude feedback. The root cause was that a few escape/proxy paths still relied on TensorToListVariable.as_proxy() self-materialization instead of the explicit materialization pattern used by tensor and torch call sites. The fix adds shared nn.Module args/kwargs materialization before proxy_args_kwargs() in lazy-module initialization, call_module, and forward-method proxy paths, and materializes lazy tolist values at generator YIELD_VALUE before storing generated_items/symbolic_result. I did not add committed tests for the suggested extra coverage because the generator and nn.Module repros already pass without this patch via later self-materialization, so they are not valid red/green regression tests under the PR requirement; I validated them with local ad hoc repros instead.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in commit 5b1c3edf4a4. Root cause: TensorToListVariable._list_variable() used a fresh copied ListVariable, so mutation methods such as append updated a disposable list while the TensorToListVariable stored in locals reconstructed from the original unmutated contents. Fix: reuse the stored materialized items for list method delegation, make length/indexing observe that materialized list, and fall back to the generic sum polyfill once materialized so post-mutation Python list semantics are preserved. I added test_tolist_list_mutation_materializes, confirmed it failed before the implementation change, and verified it passes after the fix.

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude suggestion in commit 9e5d54b983b by caching the materialized ListVariable wrapper inside TensorToListVariable and reusing it for delegated list methods and reconstruction, so mutation bookkeeping is associated with one stable wrapper instead of fresh ephemeral wrappers.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in commit ba82b279b20. Root cause: the cached materialized ListVariable wrapper is derived metadata but was not marked as a non-variable field, so traversal/clone could treat it as an independent VariableTracker subtree. Fix: added `_cached_list_var` to `TensorToListVariable._nonvar_fields` and moved the repeated lazy `tolist()` args/kwargs materialization comprehensions into one shared helper in `tensor.py`, reused by tensor, torch, and nn.Module call sites. Validation: `python3 -m py_compile torch/_dynamo/variables/tensor.py torch/_dynamo/variables/torch.py torch/_dynamo/variables/nn_module.py`, `lintrunner --take FLAKE8,RUFF ...`, and `PYTHONPATH=. python3 test/dynamo/test_misc.py -k "tolist"` (14 tests). I also inspected the current failing CI logs; the default shard failures are unrelated `build/test_custom_ops` / `libcustom_ops.so` loader failures, and the GPU torchbench failure is a SAM/CUDAGraphs accuracy failure, so I left those areas untouched.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude feedback in commit aa2baddd959.

Root cause: TensorToListVariable clone forwarded derived cache fields from __dict__ into __init__, but the constructor did not accept those cache names, so a populated lazy tolist variable could hit a TypeError on generic clone paths. The shared args/kwargs materialization helper also returned a list for varargs, which triggered the current Pyrefly lint failure.

Proposed fix: TensorToListVariable.__init__ now accepts and rebuilds the derived cache fields during clone construction, and materialize_tensor_tolist_args_kwargs now preserves varargs as a tuple. I also applied the PYFMT spacing fix in nn_module.py.

Why this is the right long-term fix: it keeps TensorToListVariable compatible with the existing VariableTracker.clone contract without preserving stale derived caches, and it makes the shared materialization helper type-compatible with all call sites that receive *args.

Validation: python3 -m py_compile torch/_dynamo/variables/tensor.py torch/_dynamo/variables/torch.py torch/_dynamo/variables/nn_module.py; PYFMT and RUFF adapters on the touched files via uv; attempted focused Dynamo tolist tests, but this checkout has no built torch shared library and stops at missing torch/lib/libtorch_global_deps.so. The remaining default-shard CI failures inspected earlier are unrelated build/test_custom_ops loader failures.

**Comment by bobrenjc93** [comment]
@claude /pr-review

**Comment by bobrenjc93** [comment]
[bot-codex-comment] Addressed the latest Claude suggestion in commit 7e82d4e1b15. Root cause: the cached ListVariable wrapper for TensorToListVariable had its own ValueMutationNew identity, while delegated list methods semantically mutate the same lazy tolist object. Proposed fix: reuse the outer TensorToListVariable mutation_type when constructing the cached ListVariable wrapper, so mutation bookkeeping observes one shared mutation identity. Why this is the right long-term fix: it keeps the materialized wrapper as an implementation detail of the lazy list while preserving the existing side-effects model if future paths track the outer variable identity. Validation: python3 -m py_compile torch/_dynamo/variables/tensor.py and git diff --check passed. Focused local Dynamo tests and lintrunner could not run in this minimal checkout because the Python environment is missing test/lint dependencies including numpy, typing_extensions, isort, and ruff. I inspected the current failed CI logs; the default-shard failures are unrelated test/custom_operator loader failures where build/test_custom_ops cannot load libcustom_ops.so, so I left that CI area untouched.

**Comment by bobrenjc93** [comment]
@claude /pr-review
